lib/level/connections.ex

```diff
@@ -3,69 +3,41 @@ defmodule Level.Connections do
   A context for exposing connections between nodes.
   """
 
+  alias Level.Pagination
+  alias Level.Spaces.Space
+  alias Level.Spaces.User
+
   @doc """
   Fetch users belonging to a space.
-
-  ## Examples
-
-      users(space, args)
-      => {:ok, %Level.Pagination.Result{
-        edges: [%User{...}],
-        page_info: %PageInfo{...},
-        total_count: 10
-      }}
-
-      users(space, invalid_args)
-      => {:error, "You cannot provide both a `before` and `after` value"}
-
   """
+  @spec users(Space.t(), map(), map()) :: {:ok, Pagination.Result.t()} | {:error, String.t()}
```

Comment:
So, typespecs!

Is there a set of keys you’re expecting the args to have? If so, you might want to define a custom type for ‘args’ to make that more clear.

Or, you might want to define that type as an alias for ‘map()’ just to make that spec clearer for folks as a fully formed idea.

Same thing with that context - if you expect it to have a shape, that’s a helpful type to define. I find types to be really helpful documentation when I’m coming on to a new project.

Defining a type like this I find to be great to give structure and names to data that might not yet be big or important enough to warrant a new struct.

That said, if things are not easily abstractable into a concrete idea, ‘map()’ will do you just fine 😀

Reply:
I'm glad you chimed in!

I will admit that I don't quite feel like I know what I'm doing yet with typespecs 😅. The docs I've found online seem a little thin on "best practices" advice. My goal is to have these serve as good documentation (and cut down on the number of verbose examples in the doc strings). I'll do some refinement here!

Reply:
I agree that best practices are still sort of emerging. If you have some time, maybe check out how we’re using them in Benchee: https://github.com/PragTob/benchee

There’s some benefit there with dialyzer, but mostly those types are documentation for us. Just giving a thing a helpful name that relates to your business logic can be super helpful, ya know?